### PR TITLE
Updates for testthat 3rd edition

### LINKED ISSUE #34: How to get rid of "tracing function "form_file" in package "curl""
Great package - I had some fears that I would be fiddling a day to get the configuration right,  things just worked out of the box!

A minor issue:  My test output is a bit polluted by this output

`tracing function "form_file" in package "curl"`

which I do not understand where it comes from. How do I suppress it?

Note: I am using the "third edition":
https://testthat.r-lib.org/articles/third-edition.html

so I cannot rule out that this is the result of the new "messages are not suppressed" feature.

```
== Testing test-AuthenticatedAccount.R ===============================================================
[ FAIL 0 | WARN 0 | SKIP 0 | PASS 0 ]Loading required package: curl
Tracing function "form_file" as seen from package "httr"
[ FAIL 0 | WARN 0 | SKIP 0 | PASS 4 ] Done!
Untracing function "form_file" in package "curl"

```

### PR DESCRIPTION
**Pull request by jonkeane**
This adjusts all of the tests to pass the 3rd edition cleanly and without excess warnings/messages. This also (fully) resolves #34 

I have punted on two tests involving `expect_failure()` within `public()`: these don't seem to trigger a failure like expected (though there are other examples where those two work just fine). We'll take this up in a followup.

### CODE REVIEW

File: .github/workflows/R-CMD-check.yaml [hunk 1]
```diff
@@ -41,6 +41,10 @@ jobs:
 
       - uses: r-lib/actions/setup-pandoc@v1
 
+      - name: System dependencies
```
Comment (nealrichardson) [inline .github/workflows/R-CMD-check.yaml:44]: If you rebase, you can remove this.

File: R/trace.R [hunk 2]
```diff
@@ -59,7 +59,7 @@ quietly <- function (expr) {
 #' @return Nothing; called for its side effects
 #' @export
 stop_mocking <- function () {
-    safe_untrace(untrace(curl::form_file))
+    safe_untrace(curl::form_file)
```
Comment (nealrichardson) [inline R/trace.R:62]: @dmenne this is probably the fix for https://github.com/nealrichardson/httptest/issues/34#issuecomment-705543832

File: tests/testthat/test-capture-requests.R [hunk 3]
```diff
@@ -73,10 +71,10 @@ test_that("We can then load the mocks it stores", {
     if (.Platform[["OS.type"]] != "windows") {
         ## Windows (on GH actions) is doing something funny here by including \r
         ## when reading from mocks (but not in the r2)
-        expect_identical(content(m2, "text"), content(r2, "text"))
+        suppressMessages(expect_identical(content(m2, "text"), content(r2, "text")))
```
Comment (nealrichardson) [inline tests/testthat/test-capture-requests.R:74]: This will conflict with master now.

Also, maybe we should add a `quiet_content()` helper function to wrap `suppressMessages(content(...))`. Basically unless the request specifies charset in the content-type header, `content()` will warn that it's assuming UTF-8. I guess the other way is to specify `content(..., encoding = "UTF-8")`, not sure if that's wise or not.

File: tests/testthat/test-expect-header.R [hunk 4]
```diff
@@ -1,20 +1,24 @@
...
                 config=add_headers(Accept="image/jpeg")),
                 "Accept: image/jpeg")))
-            expect_GET(expect_failure(expect_header(GET("http://httpbin.org/",
-                config=add_headers(Accept="image/png")),
-                "Accept: image/jpeg")))
-            expect_POST(expect_success(expect_header(POST("http://httpbin.org/",
-                config=add_headers(Accept="image/jpeg")),
-                "Accept: image/jpeg")))
-            expect_POST(expect_failure(expect_header(POST("http://httpbin.org/",
+            suppressWarnings(
```
Comment (nealrichardson) [inline tests/testthat/test-expect-header.R:7]: What's the warning that needs to be suppressed here?

File: tests/testthat/test-find-redactor.R [hunk 5]
```diff
@@ -34,7 +32,7 @@ test_that("get_current_redactor edge cases", {
 with_mock_api({
     test_that("Reading redactors from within a package (and install that package)", {
         newmocks <- tempfile()
-        expect_message(
+        suppressMessages(expect_message(
```
Comment (nealrichardson) [inline tests/testthat/test-find-redactor.R:35]: What's the deal here? `expect_message()` now expects you to assert all of the messages?

**Review by nealrichardson — COMMENTED** [review]
Thanks a lot for doing this.

File: tests/testthat/test-capture-requests.R (hunk 3 above)
Comment (jonkeane) [reply tests/testthat/test-capture-requests.R:74]: I'll merge master and unconflict it in this branch. 

Oh, I like that idea, I can definitely do that. Unless I'm missing something, I'll need to add it to the actual package and export it so that it works even inside of `public()`, right?

File: tests/testthat/test-expect-header.R (hunk 4 above)
Comment (jonkeane) [reply tests/testthat/test-expect-header.R:7]: The warning(s) are `Accept: image/png` and `Content-Type: ` (the first for the first, the second for the second and both the first and the second for the third). Which I believe [is triggered inside of `expect_header()`](https://github.com/nealrichardson/httptest/blob/master/R/expect-header.R#L29). 

If we want to catch them specifically, we can though we'll need to do it conditionally on 3e/2e since 2e suppresses them magically.

File: tests/testthat/test-find-redactor.R (hunk 5 above)
Comment (jonkeane) [reply tests/testthat/test-find-redactor.R:35]: Yeah, `expect_message()` will consume the message that it matches and the rest will percolate up. 

In {dittodb} I made a helper that runs one test in 2e and one in 3e and then do something like [the following](https://github.com/ropensci/dittodb/blob/main/tests/testthat/test-use-dittodb.R#L61-L64) where the first test is the 2e version (using `expect_message()` will match one message and eat the rest silently. The second is the 3e behavior where we can/have to for clean console output catch both of the messages.

```
  testthat_transition(
    expect_message(add_dittodb_to_helper(f), "Creating"),
    expect_message(expect_message(add_dittodb_to_helper(f), "Creating"), "Adding library")
  )
```

I'm happy to copy over `testthat_transition()` if we want to do this, but reached for `suppressMessages()` as a first step here to not add more complexity.

File: tests/testthat/test-expect-header.R (hunk 4 above)
Comment (nealrichardson) [reply tests/testthat/test-expect-header.R:7]: Right, `expect_warning()` doesn't make you assert something about *every* warning in 2e. This should be fixed in `expect_header()` one way or another, otherwise those who use it will start getting warnings in 3e.

File: tests/testthat/test-find-redactor.R (hunk 5 above)
Comment (nealrichardson) [reply tests/testthat/test-find-redactor.R:35]: Ok, here we should probably just assert all of the messages (if that's possible in both editions) since I believe we're responsible for all of the messages.

File: tests/testthat/test-use-httptest.R [hunk 6]
```diff
@@ -77,11 +75,11 @@ test_that("use_httptest integration test", {
     testpkg <- tempfile()
     dir.create(testpkg)
     expect_error(use_httptest(testpkg), "is not an R package directory")
-    
+
     desc <- file.path(testpkg, "DESCRIPTION")
     cat("Title: Foo\n", file=desc)
     helper <- file.path(testpkg, "tests", "testthat", "helper.R")
-    expect_message(use_httptest(testpkg))
+    suppressMessages(expect_message(use_httptest(testpkg)))
```
Comment (nealrichardson) [inline tests/testthat/test-use-httptest.R:82]: `suppressMessages(expect_message(...))` is odd

File: tests/testthat/test-find-redactor.R (hunk 5 above)
Comment (jonkeane) [reply tests/testthat/test-find-redactor.R:35]: ~unfortuantely~ _as it turns out_ it's not possible in both editions with a single expectation (as far as I've seen) since `expect_message()` eats all other messages  in the second edition. I can add in `testthat_transition()` and we can test both (finding one message for 2e and finding all of the message for 3e)

File: tests/testthat/test-find-redactor.R [hunk 7]
```diff
@@ -34,16 +32,34 @@ test_that("get_current_redactor edge cases", {
...
 
-                r <- GET("http://example.com/get")
-            }),
-            paste0("Using redact.R from ", dQuote("testpkg"))
+                    r <- GET("http://example.com/get")
+                }),
+                paste0("Using redact.R from ", dQuote("testpkg"))
+            ),
+            expect_message(
+                expect_message(
```
Comment (nealrichardson) [inline tests/testthat/test-find-redactor.R:49]: Hmm, there's no better way to assert multiple messages than multiple `expect_message()`s? That's too bad.

Comment (jonkeane) [reply tests/testthat/test-find-redactor.R:49]: I believe the other option would be something like [expect_snapshot()](https://testthat.r-lib.org/reference/expect_snapshot.html), but AFAICT multiple `expect_message()`s or `suppressMessages()` is [the preferred way](https://testthat.r-lib.org/reference/expect_error.html):

> New in the 3rd edition, if the code throws conditions that don't match regexp/class they will bubble up outside of the expectation, so if they're important you'll need to catch them with additional expect_message()/expect_warning() calls, or if they're unimportant you can ignore with suppressMessages()/suppressWarnings().

File: tests/testthat/test-without-internet.R [hunk 8]
```diff
@@ -49,11 +47,20 @@ public({
                 "http://httpbin.org/get",
                 '{"t')
             ## Just to be explicit since the expectations do partial matching
-            expect_failure(
-                expect_PUT(PUT("http://httpbin.org/get", body='{"test":true}'),
-                    "http://httpbin.org/get",
-                    '{"test":true}')
-            )
+            if (tryCatch(testthat::edition_get() == 3, error = function(e) FALSE)) {
```
Comment (nealrichardson) [inline tests/testthat/test-without-internet.R:50]: How about factoring this out and putting in helper.R

```
third_edition <- tryCatch(testthat::edition_get() == 3, error = function(e) FALSE)
```

And then you can write this here as `skip_if(third_edition)`

Comment (jonkeane) [reply tests/testthat/test-without-internet.R:50]: Definitely can do that — I was trying this morning to find what was going wrong with this (and failing). The call when it gets to `expect_failure()` in these two cases aren't failed somehow in 3e that I haven't tracked down yet).

I can do the refactor/skip_if() now if this would be worth merging in and following up to figure out waht is going on ^^^

File: tests/testthat/test-find-redactor.R (hunk 7 above)
Comment (nealrichardson) [reply tests/testthat/test-find-redactor.R:49]: I wonder if there's a way to write an `expect_multiple_messages()` that takes a character vector and constructs the multiple levels of `expect_message()` wrapping (on 2e it just takes the first message though). 

Maybe it's not worth the trouble but it might be nice. My only remaining concerns here are just about readability/future-maintainability of these tests.

Comment (jonkeane) [reply tests/testthat/test-find-redactor.R:49]: Yeah, absolutely. I started (and then got distracted) trying to come up with that (+submitting a PR to testthat for it), it shouldn't be hard, but getting it to work wasn't super straightforward before I got distracted.

File: tests/testthat/test-without-internet.R (hunk 8 above)
Comment (nealrichardson) [reply tests/testthat/test-without-internet.R:50]: Yeah, I think it's worth getting this mergeable and taking this issue up in a followup.

**Comment by dmenne** [comment]
Time to merge? Thanks....

File: .github/workflows/R-CMD-check.yaml [hunk 9]
```diff
@@ -41,6 +41,10 @@ jobs:
...
+        if: matrix.config.os == 'ubuntu-20.04' && matrix.config.r == '3.3'
+        run: sudo apt install libcurl4-openssl-dev
+
```
Comment (nealrichardson) [inline .github/workflows/R-CMD-check.yaml:47]:
```suggestion
```

**Comment by nealrichardson** [comment]
Yeah I think this is about ready. @jonkeane would you mind adding to NEWS about this, noting in particular the code changes (expect_header etc.) that make the package work better on 3e?

**Comment by jonkeane** [comment]
Done!

File: tests/testthat/helper.R [hunk 10]
```diff
@@ -52,3 +52,21 @@ skip_on_R_older_than <- function (version) {
         skip(paste("Requires R >=", version))
```
Comment (maelle) [inline tests/testthat/helper.R:52]: this file should be called setup.R https://testthat.r-lib.org/reference/test_dir.html#special-files

**Comment by maelle** [comment]
I have another unsollicited comment, not related to this PR in particular: it seems several tests fail with `devtools::test()`. Is it a known problem? Would there be a way around this?

**Comment by nealrichardson** [comment]
@maelle I'm not aware of a general problem; #15 mentions devtools::test() but it's a bit murky as to what is going on. Could you please open an issue with details about that?

File: tests/testthat/helper.R (hunk 10 above)
Comment (maelle) [reply tests/testthat/helper.R:52]: Yes sorry

**Comment by maelle** [comment]
So when you run `devtools::test()` locally on httptest all tests pass? I'll open an issue tomorrow or on Thursday

**Comment by nealrichardson** [comment]
I don't personally do `devtools::test()` to run the tests, I always run in a fresh R process

**Comment by nealrichardson** [comment]
Ok, I squashed/rebased this on master after merging @maelle's latest PR, which conflicted with this. I think it's good now and will merge after CI passes, though I'll take one last pass through to make sure I didn't screw up the git shenanigans.
